lib/key/check.js

```diff
@@ -1,6 +1,102 @@
 import { openpgp } from '../openpgp';
 import { serverTime } from '../serverTime';
 
+/**
+ * Check the strength of the (sub)key based on the given requirements.
+ * @param {OpenPGPKey|openpgp.SubKey} publicKeyOrSubkey - (sub)key to check
+ * @param {Set<openpgp.enums.publicKey} requirements.allowedPublicKeyAlgorithms
+ * @param {Set<openpgp.enums.curve} requirements.allowedCurves
+ * @throws {Error} if the kay is considered too weak
+ */
+function checkKeyStrength(publicKeyOrSubkey, requirements) {
+    const { allowedCurves, allowedPublicKeyAlgorithms, minRSABits } = requirements;
+    const keyInfo = publicKeyOrSubkey.keyPacket.getAlgorithmInfo();
+    const keyAlgo = openpgp.enums.write(openpgp.enums.publicKey, keyInfo.algorithm);
+
+    if (!allowedPublicKeyAlgorithms.has(keyAlgo)) {
+        throw new Error(`${keyInfo.algorithm} keys are considered unsafe`);
+    }
+
+    if (keyInfo.curve && !allowedCurves.has(keyInfo.curve)) {
+        throw new Error(`Keys using curve ${keyInfo.curve} are considered unsafe`);
+    }
+
+    if (keyInfo.rsaBits && keyInfo.rsaBits < minRSABits) {
+        throw new Error(`Keys shorter than ${minRSABits} bits are considered unsafe`);
+    }
+}
+
+/**
+ * Checks whether the primary key and the encryption subkey are strong enough to be imported as personal keys.
+ * These checks are stronger than those for contact keys.
+ * @param {OpenPGPKey} publicKey - key to check
+ * @throws {Error} if the key is considered too weak
+ * @async
+ */
+export async function checkPersonalKeyStrength(publicKey) {
+    const { enums } = openpgp;
+    // primary key
+    checkKeyStrength(publicKey, {
+        allowedPublicKeyAlgorithms: new Set([
+            enums.publicKey.rsa_encrypt_sign,
+            enums.publicKey.rsa_sign,
+            enums.publicKey.ecdsa,
+            enums.publicKey.eddsa
+        ]),
+        allowedCurves: new Set([enums.curve.ed25519, enums.curve.p256, enums.curve.p384, enums.curve.p521]),
+        minRSABits: 2048
+    });
+    // TODO should we check all subkeys instead? (faster, but we dont know if they are signing or encryption subkeys)
```

Comment:
?

Reply:
Yes, let's just check all subkeys, and say that all of them must be RSA/ECDSA/EdDSA/ECDH (without particularly looking at whether it's a signing or encryption key)